src/jdk.jartool/share/classes/sun/tools/jar/Validator.java

```diff
@@ -62,20 +93,188 @@ final class Validator {
     private Set<String> concealedPkgs = Collections.emptySet();
     private ModuleDescriptor md;
     private String mdName;
+    private final ZipInputStream zis;
 
-    private Validator(Main main, ZipFile zf) {
+    private Validator(Main main, ZipFile zf, ZipInputStream zis) {
         this.main = main;
         this.zf = zf;
+        this.zis = zis;
         checkModuleDescriptor(MODULE_INFO);
     }
 
-    static boolean validate(Main main, ZipFile zf) throws IOException {
-        return new Validator(main, zf).validate();
+    static boolean validate(Main main, File zipFile) throws IOException {
+        try (ZipFile zf = new ZipFile(zipFile);
+            ZipInputStream zis = new ZipInputStream(new BufferedInputStream(
+                    new FileInputStream(zipFile)))) {
+            return new Validator(main, zf, zis).validate();
+        }
+    }
+
+    /**
+     * Validate that the CEN/LOC file name header field adheres  to
+     * PKWARE APPNOTE-6.3.3.TXT:
+     *
+     * 4.4.17.1 The name of the file, with optional relative path.
+     * The path stored MUST not contain a drive or
+     * device letter, or a leading slash.  All slashes
+     * MUST be forward slashes '/' as opposed to
+     * backwards slashes '\' for compatibility with Amiga
+     * and UNIX file systems etc.
+     * Also validate that the file name is not "." or "..", and that any name
+     * element is not equal to "." or ".."
+     *
+     * @param entryName ZIP entry name
+     * @return true if a valid Zip Entry file name; false otherwise
+     */
+    public static boolean isZipEntryNameValid(String entryName) {
+        return !INVALID_ZIP_ENTRY_NAME_PATTERN.matcher(entryName).find();
+    }
+
+    /**
+     * Validate base on entries in CEN and LOC. To ensure
+     * - Valid entry name
+     * - No duplicate entries
+     * - CEN and LOC should have same entries, in the same order
+     *
+     * NOTE: In order to check the encounter order based on the CEN listing,
+     *       this implementation assumes CEN entries are to be added before
+     *       add any LOC entries. That is, addCenEntry should be called before
+     *       calls to addLocEntry to ensure encounter order can be compared
+     *       properly.
+     */
+    private class EntryValidator {
+        // A place holder when an entry is not yet seen in the directory
+        static EntryEncounter PLACE_HOLDER = new EntryEncounter(0, 0);
```

Comment:
It looks like this could be `final`? And then `isPlaceHolder()` could just return `this == PLACE_HOLDER`?